repos/system_upgrade/el7toel8/actors/cupsmigrate/libraries/cupsmigrate.py

```diff
@@ -0,0 +1,232 @@
+
+    # we need to decide whether we copy the files to /etc/cups/ssl
+    # or set ServerKeychain to non-default directory or do nothing
+    if all(value.rsplit('/', 1)[0] == '/etc/cups/ssl'
+           for value in certkey_values):
+        return
+
+    if all(v.rsplit('/', 1)[0] == certkey_values[0].rsplit('/', 1)[0]
+           for v in certkey_values):
+        path = certkey_values[0].rsplit('/', 1)[0]
+        op.write(CUPSFILES_CONF, 'a',
+                 '\n# added by Leapp\nServerKeychain {}\n'.format(path))
+    else:
+        for value in certkey_values:
+            if not value.rsplit('/', 1)[0] == '/etc/cups/ssl':
+                op.copy_to_ssl(value)
+
+
+def _get_facts(model):
+    """
+    Consumes input data model
+
+    :param class model: name of model which we consume
+    """
+    return next(api.consume(model), None)
+
+
+def migrate_configuration(error_log=api.current_logger().error,
+                          debug_log=api.current_logger().debug,
+                          op=FileOperations(),
+                          consume_function=_get_facts):
+    """
+    Migrate CUPS configuration based on gathered facts
+
+    :param func error_log: sends error messages
+    :param func debug_log: sends debug messages
+    :param obj op: IO operations
+    :param func consume_function: receives data object from a model
+    """
+
+    facts = consume_function(CupsChangedFeatures)
+    error_list = []
+
+    if not facts:
+        return
+
+    if facts.include:
+        debug_log('Migrating CUPS configuration - Include directives.')
+        include_errors = []
+
+        include_errors = migrate_include(facts.include_files, op)
+        if include_errors:
+            error_list += include_errors
+
+    if facts.digest:
+        debug_log('Migrating CUPS configuration - BasicDigest/Digest directives.')
+
+        try:
+            migrate_digest(op)
+        except IOError as error:
+            error_list.append('Digest/BasicDigest values: {}'.format(error))
+
+    if facts.env:
+        debug_log('Migrating CUPS configuration - PassEnv/SetEnv directives.')
+
+        try:
+            move_directives(['PassEnv', 'SetEnv'],
+                            op)
+        except IOError as error:
+            error_list.append('PassEnv/SetEnv directives: {}'.format(error))
+
+    if facts.certkey:
+        debug_log('Migrating CUPS configuration - '
+                  'ServerKey/ServerCertificate directive.')
+
+        try:
+            migrate_certkey(op)
+        except IOError as error:
+            error_list.append('ServerKey/ServerCertificate directives: {}'.format(error))
+
+    if facts.printcap:
+        debug_log('Migrating CUPS configuration - PrintcapFormat directive.')
+
+        try:
+            move_directives(['PrintcapFormat'],
+                            op)
+        except IOError as error:
+            error_list.append('PrintcapFormat directive: {}'.format(error))
+
+    if error_list:
```

Comment:
@zdohnal we should ad here `TODO: add post upgrade report`. As we are not doing post-upgrade reports so much yet, we do not have to add it now. But we should add least add TODO here.